File: tree/elementsAtdistD.cpp
#include<iostream>
#include<vector>
#include<cmath> 
#include<algorithm>
#include<set>
#include<string>
#include<map>
#include<cstring>
#include<stack>
#include<queue>

using namespace std;

#define ll long long

const int INF = 1e9+7;
const int mod = 1e9+7;
const int MAX = 1e5+5;

vector<int> tree[MAX];
int level[MAX];
int vis[MAX];

int n, dis;

void dfs(int s){
	vis[s] = 1;
	for(auto to : tree[s]){
		if(!vis[to]){
			level[to] = level[s]+1; // increasing distance level by level
			if(level[to] == dis){
				cout << to << " "; // printing elements at distance D
			}
			dfs(to);
		}
	}
}

int main(){
	//freopen("tree.in","r",stdin);
    //freopen("tree.out","w",stdout);
    //freopen("input.txt", "r", stdin);
    //freopen("out.txt", "w", stdout);
    ios_base::sync_with_stdio(0);
    cin.tie(0);

    cin >> n >> dis;

    for(int i = 1; i < n; ++i){
    	int a, b;
    	cin >> a >> b;
    	tree[a].push_back(b);
    	tree[b].push_back(a);
    }

    level[2] = 1; // taking root as 2
    dfs(2);
    cout << endl;
}

File: tree/elementsAtdistD.py
from collections import defaultdict

INF =  (10**5)+5
vis = [0]*INF
level = [0]*INF
dp = [0]*INF

tree = defaultdict(list)

def dfs(s):
	vis[s] = 1
	for i in tree[s]:
		if(not vis[i]):
			level[i] = level[s]+1
			if(level[i] == dist):
				print(i, end = " ")
			dfs(i)


if __name__ == '__main__':
	n, dist = [int(i) for i in input().split()]
	for i in range(1, n):
		u, v = [int(i) for i in input().split()]
		tree[u].append(v)
		tree[v].append(u)

	level[1] = 1 # level 1 at root 1
	dfs(1)
	print("")	


